### ISSUE TITLE
Tokenizer does not support context sensitive parsing

### ISSUE THREAD

**Issue by kukulich** [squizlabs/PHP_CodeSniffer#832]
In PHP7 works this code thanks to context sensitive parser:

```php
class Foo
{
	const PRIVATE = 'private';
}
```

However it reports errors in PHP_CodeSniffer because `token_get_all()` supports context sensitive parsing only with `TOKEN_PARSE` flag, see https://bugs.php.net/bug.php?id=71207.

I would like to prepare pull request but there's a problem that `token_get_all()` with `TOKEN_PARSE` flag cannot parse patters in `PEAR_Sniffs_ControlStructures_ControlSignatureSniff` or `Squiz_Sniffs_Functions_FunctionDeclarationSniff`, it throws a parse error. I would suggest to replace patterns  with the output of `token_get_all()` so the tokens would be hardcoded in the sniffs but that's probably a BC break. Do you have another suggestion?

I have prepared temporary fix: https://github.com/slevomat/PHP_CodeSniffer/commit/3cc858155f2857d4276ecf213cc43ca075b59898

**Comment by gsherwood** [comment]
The PHPCS tokenizer already converts quite a lot of tokens based on context, and adds a lot that PHP has left out. So I think the best thing to do is just find and fix the tokens manually as it does for the others.

Here is a recent example where class names of `True`, `False`, and `Null` were being tokenized as true/false/null when they should have just been strings: https://github.com/squizlabs/PHP_CodeSniffer/commit/f9e4b2d40272e33ccd3b8544e1e7ef4d782354df

So I'm thinking something like this is easy enough, and allows for parsing PHP7 code in older PHP versions.

**Comment by gsherwood** [comment]
Marking this as a bug because the actual problem is with errors being reported when they shouldn't. Fixing it in the tokenizer should fix pretty much everything.

**Comment by kukulich** [comment]
Ok, I look at the tokenizer.

**Comment by kukulich** [comment]
@gsherwood I've prepared pull request: https://github.com/squizlabs/PHP_CodeSniffer/pull/836

**Comment by gsherwood** [comment]
Thanks a lot for the PR to fix this issue.
